3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1535 @@
+Afterwards we will split things up and go into more details per each
+individual section.
+
+```json
+{
+    "maxGSS": 1,
+    "vertexAccumulators": {
+      "outDegree": {
+        "accumulatorType": "store",
+        "valueType": "ints"
+      },
+      "inDegree": {
+        "accumulatorType": "store",
+        "valueType": "ints"
+      }
+    },
+    "phases": [{
+      "name": "main",
+      "initProgram": ["seq",
+        ["accum-set!", "outDegree", ["this-outbound-edges-count"]],
+        ["accum-set!", "inDegree", 0],
+        ["send-to-all-neighbours", "inDegree", 1]
+      ],
+      "updateProgram": ["seq",
+        false]
+    }],
+    "dataAccess": {
+      "writeVertex": [
+        "attrib-set", ["attrib-set", ["dict"], "inDegree", ["accum-ref", "inDegree"]],
+        "outDegree", ["accum-ref", "outDegree"]
+      ]
+    }
+  }
+```
+
+#### Used Accumulators
+
+In the example, we are using two vertex accumulators: `outDegree` and `inDegree`,
+as we want to calculate and store two values.
+
+##### outDegree
+
+```json
+"outDegree": {
+  "accumulatorType": "store",
+  "valueType": "ints"
+}
+```
+
+A vertex knows exactly how many outgoing edges it has by definition. Therefore
+we only have to set the amount to an accumulator once and not multiple times.
+With that knowledge it makes sense to set the `accumulatorType` to store, as
+no further calculations need to take place. As the possible amount of
+outgoing edges is even, we are setting `valueType` to `ints`. Additionally, we
+do not need to store the sender as that value is out of interest in that example.
+
+##### inDegree
+
+What a vertex does **not know** is, how many incoming edges it has.
+Therefore we need to get them to know that value.
+
+```json
+"inDegree": {
+  "accumulatorType": "sum",
+  "valueType": "ints"
+}
+```
+
+The choice for `valueType` is equal compared to "outDegree"
+because of the same reason. But as you can see, the `accumulatorType` is now
+set to `sum`. As our vertices do not know how many incoming edges there are,
+each vertex needs to send a message to all outgoing. In our program, a message
+will be sent to every neighbor. That means a vertex with **n** neighbors,
+will send **n** messages. As in our case, a single message represents a single
+incoming edge, we need to add **"+1"** to our accumulator per each message, and
+therefor set the `accumulatorType` to `sum`.
+
+#### Program
+
+##### initProgram
+
+```js
+initProgram: [
+  "seq",
+
+  ["accum-set!", "outDegree", ["this-outbound-edges-count"]], // Sets our outDegree accumulator with ["this-outbound-edges-count"]
+
+  // Init in degree to 0
+  ["accum-set!", "inDegree", 0],             // Initializes our inDegree (sum) accumulator to 0
+  ["send-to-all-neighbours", "inDegree", 1]  // Sends value: "1" to all neighbors, so their inDegree can be raised next round!
+]
+```
+
+##### updateProgram
+
+```
+updateProgram: ["seq",
+  false]
```

Comment:
Maybe 
```suggestion
  "vote-halt"]
```
would showcase this better?

Reply:
You could even just write `update-program: "vote-halt"`